dyno-jedis/src/main/java/com/netflix/dyno/jedis/DynoJedisClient.java

```diff
@@ -4060,6 +4066,530 @@ public long pfcount(String... keys) {
+            Logger.debug("Expire hash:{} inconsistent with metadata:{}. Failed to delete expired fields from hash",
+                    ehMetadataUpdateResult.dataKey, ehMetadataUpdateResult.metadataKey);
+        }
+
+        // if fields were not deleted from metadata, correctness is not affected.
+        if (ehMetadataUpdateResult.expiredFields.size() > 0 && ehMetadataUpdateResult.zremResponse != null &&
+                ehMetadataUpdateResult.expiredFields.size() != ehMetadataUpdateResult.zremResponse.get()) {
+            Logger.debug("Expire hash:{} inconsistent with metadata:{}. Failed to delete expired fields from metadata",
+                    ehMetadataUpdateResult.dataKey, ehMetadataUpdateResult.metadataKey);
+        }
+    }
+
+    @Override
+    public Long ehset(final String key, final String field, final String value, final long ttl)
+            throws UnsupportedOperationException, DynoException {
+        final DynoJedisPipeline pipeline = this.pipelined();
+        final Response<Long> zResponse = pipeline.zadd(ehashMetadataKey(key), timeInEpochSeconds(ttl), field,
+                ZAddParams.zAddParams().ch());
+        final Response<Long> hResponse = pipeline.hset(ehashDataKey(key), field, value);
+        pipeline.sync();
+
+        // If metadata operation failed, remove the data and throw exception
+        if (zResponse.get() != 1) {
+            d_hdel(ehashDataKey(key), field);
+            throw new DynoException("Failed to set field:" + field + "; Metadata operation failed:" + zResponse.get());
+        }
+        return hResponse.get();
+    }
+
+    public Long ehsetnx(final String key, final String field, final String value, final long ttl) {
+        final String ehashDataKey = ehashDataKey(key);
+        final String ehashMetadataKey = ehashMetadataKey(key);
+        final DynoJedisPipeline pipeline = this.pipelined();
+        final Response<Long> zResponse = pipeline.zadd(ehashMetadataKey, timeInEpochSeconds(ttl), field,
+                ZAddParams.zAddParams().ch());
+        final Response<Long> hResponse = pipeline.hsetnx(ehashDataKey, field, value);
+        pipeline.sync();
+
+        // If metadata operation failed, remove the data and throw exception
+        if (zResponse.get() != hResponse.get()) {
+            d_hdel(ehashDataKey, field);
+            d_zrem(ehashMetadataKey, field);
+            throw new DynoException("Metadata inconsistent with data for expireHash: " + ehashDataKey);
+        }
+        return hResponse.get();
+    }
+
+    @Override
+    public String ehget(final String key, final String field)
+            throws UnsupportedOperationException, DynoException {
+        final String dataKey = ehashDataKey(key);
+        final DynoJedisPipeline pipeline = this.pipelined();
+
+        EHMetadataUpdateResult ehMetadataUpdateResult = ehPurgeExpiredFields(pipeline, key);
```

Comment:
It seems like a lot of functions have very similar steps. To avoid code duplication, is it possible to create a template and call that template once per function?

Eg: Have a more generic function like ehTask() or something:
ehTask(pipeline, key, field, func_ptr);

'func_ptr' would be 'hget', or 'hexists', etc.

And the last error checking step would be generic to all types of queries.

What do you think? This can be added as a future patch too.